### ISSUE TITLE
debug build causes `BN_GF2m_mod_inv` result differences

### ISSUE THREAD

**Issue by gal1ium** [openssl/openssl#19826]
Hi! There is a weird issue we met in the bignum library from the master branch. The following program in **debug build** would generate different random values every time, so I think there might be some memory corruptions inside.
```
#include <openssl/bn.h>
BIGNUM* r0;
BIGNUM* r1;
BIGNUM* r2;
BIGNUM* r3;

int main(void)
{
  r0 = BN_dup(BN_get0_nist_prime_256());
  r1 = BN_dup(BN_get0_nist_prime_521());
  r2 = BN_new();
  r3 = BN_dup(BN_get0_nist_prime_224());
  BN_lshift1(r2, r3);
  BN_GF2m_mod_inv(r0, r1, r2, BN_CTX_new());
  printf("hex:%s\n", BN_bn2hex(r0));
  return 0;
}
```

while the following similar program without those `BN_dup` seems to work normally:
```
#include <openssl/bn.h>
BIGNUM* r0;
BIGNUM* r1;
BIGNUM* r2;
BIGNUM* r3;

int main(void)
{
  r0 = BN_get0_nist_prime_256();
  r1 = BN_get0_nist_prime_521();
  r2 = BN_new();
  r3 = BN_get0_nist_prime_224();
  BN_lshift1(r2, r3);
  BN_GF2m_mod_inv(r0, r1, r2, BN_CTX_new());
  printf("hex:%s\n", BN_bn2hex(r0));
  return 0;
}
```

with a little debugging we found that it might be due to the debug flag `BN_FLG_FIXED_TOP`

**Comment by t8m** [comment]
Have you checked the return value from BN_GF2m_mod_inv()?

These functions are not supposed to work with random input values but with binary field polynomials converted into the BIGNUM represetnation.

**Comment by gal1ium** [comment]
> Have you checked the return value from BN_GF2m_mod_inv()?
> 
> These functions are not supposed to work with random input values but with binary field polynomials converted into the BIGNUM representation.

Ah, got it, seems like we were messing up the inputs unintendedly by design. 
And later I also got an ASAN report like this:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1494861==ERROR: AddressSanitizer: SEGV on unknown address 0x607ff7d7d8b8 (pc 0x0000004c801d bp 0x6080000000a0 sp 0x7fffffffe340 T0)
==1494861==The signal is caused by a WRITE memory access.
    #0 0x4c801d in BN_GF2m_mod_arr ~/openssl/crypto/bn/bn_gf2m.c:336:22
    #1 0x4c8ea3 in BN_GF2m_mod_mul_arr ~/openssl/crypto/bn/bn_gf2m.c:449:9
    #2 0x4c9c70 in BN_GF2m_mod_mul ~/openssl/crypto/bn/bn_gf2m.c:484:11
    #3 0x4ca0b6 in BN_GF2m_mod_inv ~/openssl/crypto/bn/bn_gf2m.c:746:10
    #4 0x4c5124 in main (~/a.out+0x4c5124)
    #5 0x7ffff7c45082 in __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:308:16
    #6 0x41d04d in _start (~/a.out+0x41d04d)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV ~/openssl/crypto/bn/bn_gf2m.c:336:22 in BN_GF2m_mod_arr
==1494861==ABORTING
```
maybe it could be verified whether there is any memory issue to be fixed?

**Comment by gal1ium** [comment]
> What is the reproducer for the ASAN failure?

It's still this unintended usage:
```
#include <openssl/bn.h>
BIGNUM* r0;
BIGNUM* r1;
BIGNUM* r2;
BIGNUM* r3;

int main(void)
{
  r0 = BN_dup(BN_get0_nist_prime_256());
  r1 = BN_dup(BN_get0_nist_prime_521());
  r2 = BN_new();
  r3 = BN_dup(BN_get0_nist_prime_224());
  BN_lshift1(r2, r3);
  BN_GF2m_mod_inv(r0, r1, r2, BN_CTX_new());
  printf("hex:%s\n", BN_bn2hex(r0));
  return 0;
}
```
but might expose some memory issues, on the master branch

**Comment by t8m** [comment]
I do not think it makes sense to call these functions with random BIGNUMs as parameters. Unless there is a concrete example where the overflow happens with at least somewhat legitimate inputs, I do not think it makes sense to investigate further.

**Comment by dfabregat** [comment]
Hi guys. I'm new around here, but if I may... I took a look at the asan issue and this is what I'm observing.

TL;DR: I believe the preconditions of `BN_GF2m_mod_inv` are being violated by passing an even modulus (`r2 = r3 < 1`). Thus, no bug to fix, just undefined behavior.

Now to the details:

The code by @gal1ium  calls `BN_GF2m_mod_inv` with an even modulus (`r2 = r3 < 1`). This code will end up calling `BN_GF2m_poly2arr` with that same modulus, and the resulting array is passed down to `BN_GF2m_mod_arr`.

Calling these functions with an even modulus violates their preconditions  since, according to the [changelog for 1.1.1](https://www.openssl.org/news/cl111.txt), the coefficient of t^0 (or LSB of `r2` in this case) should be set:

     For some functions, an the irreducible polynomial defining a
     field can be given as an 'unsigned int[]' with strictly
     decreasing elements giving the indices of those bits that are set;
     i.e., p[] represents the polynomial
          f(t) = t^p[0] + t^p[1] + ... + t^p[k]
     where
          p[0] > p[1] > ... > p[k] = 0.
     This applies to the following functions:

          BN_GF2m_mod_arr
          BN_GF2m_mod_mul_arr
          BN_GF2m_mod_sqr_arr
          BN_GF2m_mod_inv_arr        [wrapper for BN_GF2m_mod_inv]
          BN_GF2m_mod_div_arr        [wrapper for BN_GF2m_mod_div]
          BN_GF2m_mod_exp_arr
          BN_GF2m_mod_sqrt_arr
          BN_GF2m_mod_solve_quad_arr
          BN_GF2m_poly2arr
          BN_GF2m_arr2poly

What happens in the code is that `BN_GF2m_poly2arr` generates the following array:

```
p[0] = 224
p[1] = 223
...
p[127] = 97
p[128] =  1
p[129] = -1
```

leading to the error caught by asan later on in `bn_gf2m.c:336`, since the loop in `bn_gf2m.c:330` does not stop when expected (`p[k] != 0`) and the code finds garbage beyond `p[129] = -1`.

**Comment by mattcaswell** [comment]
Closing since this is marked as resolved.
